packages/core/src/indexing/index-api.ts

```diff
@@ -4,14 +4,47 @@ import type {
   Page,
   Pagination,
   StreamState,
+  DiagnosticsLogger,
 } from '@ceramicnetwork/common'
+import { SyncOptions } from '@ceramicnetwork/common'
 import type { DatabaseIndexApi } from './database-index-api.js'
-import { NotImplementedError } from './not-implemented-error.js'
+import type { Repository } from '../state-management/repository.js'
 
+/**
+ * API to query an index.
+ */
 export class IndexApi implements IndexClientApi {
-  constructor(private readonly databaseIndexApi: DatabaseIndexApi | undefined) {}
+  constructor(
+    private readonly databaseIndexApi: DatabaseIndexApi | undefined,
+    private readonly repository: Repository,
+    private readonly logger: DiagnosticsLogger
+  ) {}
 
-  queryIndex(query: BaseQuery & Pagination): Promise<Page<StreamState>> {
-    throw new NotImplementedError('IndexApi::queryIndex')
+  /**
+   * Query the index. Ask an indexing database for a list of StreamIDs,
+   * and convert them to corresponding StreamState instances via `Repository::streamState`.
+   *
+   * We assume that a state store always contains StreamState for an indexed stream.
+   */
+  async queryIndex(query: BaseQuery & Pagination): Promise<Page<StreamState>> {
```

Comment:
I'm also worried that we don't have any protection against a client sending a query requesting a huge number of results in a single batch.  I think we should have some server-enforced maximum batch size. I'll file a ticket for this.